pallets/orml-currencies-allowance-extension/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+		<T as frame_system::Config>::AccountId,
+	>>::Balance;
+
+pub(crate) type CurrencyOf<T> =
+	<<T as orml_currencies::Config>::MultiCurrency as orml_traits::MultiCurrency<
+		<T as frame_system::Config>::AccountId,
+	>>::CurrencyId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{pallet_prelude::*, transactional};
+	use frame_system::{ensure_root, pallet_prelude::OriginFor};
+
+	use super::*;
+
+	/// ## Configuration
+	/// The pallet's configuration trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + orml_tokens::Config + orml_currencies::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		AllowedCurrenciesAdded {
+			currencies: Vec<CurrencyOf<T>>,
+		},
+		AllowedCurrenciesDeleted {
+			currencies: Vec<CurrencyOf<T>>,
+		},
+		/// (Additional) funds have been approved for transfer to a destination account.
+		TransferApproved {
+			currency_id: CurrencyOf<T>,
+			source: T::AccountId,
+			delegate: T::AccountId,
+			amount: BalanceOf<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Parachain is not running.
+		Unapproved,
+		ParachainNotRunning,
+		CurrencyNotLive,
+	}
+
+	/// Approved balance transfers. Balance is the amount approved for transfer.
+	/// First key is the currency ID, second key is the owner and third key is the delegate.
+	#[pallet::storage]
+	pub type Approvals<T: Config> = StorageNMap<
+		_,
+		(
+			NMapKey<Blake2_128Concat, CurrencyOf<T>>,
+			NMapKey<Blake2_128Concat, T::AccountId>, // owner
+			NMapKey<Blake2_128Concat, T::AccountId>, // delegate
+		),
+		BalanceOf<T>,
+	>;
+
+	/// Currencies that can be used in chain extension
+	#[pallet::storage]
+	pub(super) type AllowedCurrencies<T: Config> =
+		StorageMap<_, Blake2_128Concat, CurrencyOf<T>, ()>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub allowed_currencies: Vec<CurrencyOf<T>>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { allowed_currencies: vec![] }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			for i in &self.allowed_currencies.clone() {
+				AllowedCurrencies::<T>::insert(i, ());
+			}
+		}
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	// The pallet's dispatchable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Added allowed currencies that possible to use chain extension
+		///
+		/// # Arguments
+		/// * `currencies` - list of currency id allowed to use in chain extension
+		#[pallet::call_index(1)]
+		#[pallet::weight(10000)]
```

Comment:
Should we not use actual benchmarking to determine the weights?

Reply:
Also for the other extrinsics here.

Reply:
> Should we not use actual benchmarking to determine the weights?

oh. sure. good catch. will update. thanks 👍

Reply:
I already mentioned this [here](https://github.com/pendulum-chain/pendulum/pull/163#discussion_r1138903838) 😶‍🌫️

Reply:
Done.